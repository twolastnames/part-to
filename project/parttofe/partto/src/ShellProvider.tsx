import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
