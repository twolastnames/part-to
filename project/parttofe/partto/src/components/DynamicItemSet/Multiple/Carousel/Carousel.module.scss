.multiple {
  display: flex;
  height: 100%;
}

.carouselButton {
  margin: 0 50%;
  flex: 0 1 auto;
}

.carousel {
  height: 100%;
  width: 100%;
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;
}

.content {
  flex: 1 1 auto;
}

.slide {
  height: 100%;
  width: 100%;
}

.buttonSet {
  flex: 1 0 auto;
  display: flex;
  flex-direction: column;
  height: 100%;
}

.buttons {
  justify-content: center;
  flex: 1 0 auto;
}
