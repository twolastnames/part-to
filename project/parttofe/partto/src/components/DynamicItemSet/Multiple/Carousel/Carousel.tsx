import React, { ReactNode, useEffect, useState } from "react";

import classes from "./Carousel.module.scss";
import { Button, ButtonProps } from "../../../Button/Button";
import { Down, Up } from "../../../Icon/Icon";
import { ButtonSet } from "../../ButtonSet/ButtonSet";
import {
  IterationProgress,
  marks,
} from "../IterationProgress/IterationProgress";
import { MultipleProps } from "../Multiple";
import { Item } from "../../DynamicItemSet";

export type CarouselPage = { view: ReactNode; operations?: Array<ButtonProps> };

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: number) => number) => void,
) => void;

const goBack: GoADirection = (length, setSelected) => {
  setSelected((last: number) => (last <= 0 ? length - 1 : last - 1));
};
const goForward: GoADirection = (length, setSelected) => {
  setSelected((last) => (last >= length - 1 ? 0 : last + 1));
};

function getPages(current: Array<Item>) {
  return current.map(({ key, detailView, itemOperations }) => ({
    key,
    view: detailView,
    operations: itemOperations,
  }));
}

export function Carousel({ items }: MultipleProps) {
  const [showDuration, setShowDuration] = useState<number>(5);
  const [selected, setSelected] = useState<number>(0);
  const [pages, setPages] = useState<Array<CarouselPage & { key: string }>>(
    getPages(items),
  );

  useEffect(() => {
    if (marks[showDuration]?.duration == null) {
      return;
    }
    const id = setInterval(
      () => {
        goForward(items.length, setSelected);
      },
      (marks[showDuration] == null
        ? marks[2]
        : marks[showDuration]
      ).duration.toMilliseconds(),
    );
    setPages(getPages(items));
    return () => {
      clearInterval(id);
    };
  }, [showDuration, items]);

  const operations = pages[selected]?.operations;
  return (
    <div
      key={pages[selected]?.key}
      className={classes.multiple}
      data-testid="Carousel"
    >
      <div className={classes.carousel}>
        <div className={classes.carouselButton}>
          <Button
            icon={Up}
            text="Back One"
            onClick={() => {
              goBack(pages.length, setSelected);
            }}
          />
        </div>
        <div className={classes.content}>
          <div key={pages[selected]?.key} className={classes.listView}>
            {pages[selected]?.view}
          </div>
        </div>
        <div className={classes.carouselButton}>
          <Button
            icon={Down}
            text="Forward One"
            onClick={() => {
              goForward(pages.length, setSelected);
            }}
          />
        </div>
      </div>
      <IterationProgress
        setShowDuration={setShowDuration}
        on={selected}
        total={pages.length}
        showDuration={showDuration}
      />
      {operations ? (
        <div className={classes.buttonSet}>
          <div className={classes.buttons}>
            <ButtonSet operations={operations} />
          </div>
        </div>
      ) : (
        <></>
      )}
    </div>
  );
}
