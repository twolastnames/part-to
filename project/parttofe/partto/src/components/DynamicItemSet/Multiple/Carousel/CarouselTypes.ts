import { ReactNode } from "react";
import { ButtonProps } from "../../../Button/ButtonTypes";

export type CarouselPage = { view: ReactNode; operations?: Array<ButtonProps> };
