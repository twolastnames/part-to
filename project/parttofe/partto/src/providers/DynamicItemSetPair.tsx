import React, {
  createContext,
  useState,
  Context,
  MutableRefObject,
  createRef,
  useContext,
  SetStateAction,
} from "react";
import { PropsWithChildren } from "react";
import { Duration, getDuration } from "../shared/duration";
import { debounce, DebouncedFunc } from "lodash";

export type DynamicItemSetPairType = {
  selected: number;
  setSelected: (selected: number) => void;
  getTotal: () => number;
  goBack: () => void;
  goForward: () => void;
  showDuration: Duration;
  togglePause: () => void;
  paused: boolean;
  nextShowDuration: () => void;
};

export const UndefinedDynamicItemSetPair = {
  selected: 0,
  setSelected: (_: number) => undefined,
  goBack: () => undefined,
  goForward: () => undefined,
  getTotal: () => 0,
  showDuration: getDuration(0),
  togglePause: () => undefined,
  paused: true,
  nextShowDuration: () => undefined,
};

export function useSingleOfPair({ context }: ContextDescription) {
  return useContext(context) || UndefinedDynamicItemSetPair;
}

type ContextType = Context<undefined | DynamicItemSetPairType>;

export type ContextDescription = {
  context: ContextType;
  setCount: (value: number) => void;
  setSettingKey: (settingKey: string) => void;
  setDurations: (value: Array<Duration>) => void;
};

const defaultDurations = [1000, 2000, 4000, 8000];

const getDurations = (key: string) =>
  JSON.parse(
    localStorage.getItem(`${key}Durations`) || JSON.stringify(defaultDurations),
  ).map(getDuration);

const setDurations = (key: string, value: Array<Duration>) => {
  localStorage.setItem(
    `${key}Durations`,
    JSON.stringify(value.map((duration) => duration.toMilliseconds())),
  );
};

const leftCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const leftSettingKey: MutableRefObject<string | null> = createRef();
const getLeftSettingKey = () => leftSettingKey.current || "left";
function setLeftCount(value: number) {
  leftCount.current && leftCount.current(value);
}
const LeftDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const LeftContext: ContextDescription = {
  context: LeftDynamicItemSetPairContext,
  setCount: setLeftCount,
  setSettingKey: (value: string) => {
    leftSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getLeftSettingKey(), value);
  },
};

const rightCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const rightSettingKey: MutableRefObject<string | null> = createRef();
const getRightSettingKey = () => leftSettingKey.current || "left";

function setRightCount(value: number) {
  rightCount.current && rightCount.current(value);
}
const RightDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const RightContext: ContextDescription = {
  context: RightDynamicItemSetPairContext,
  setCount: setRightCount,
  setSettingKey: (value: string) => {
    rightSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getRightSettingKey(), value);
  },
};

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: ProviderState) => ProviderState) => void,
) => void;

const goBack: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected <= 0 ? length - 1 : last.selected - 1,
  }));
};

const goForward: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected >= length - 1 ? 0 : last.selected + 1,
  }));
};

type ProviderState = {
  selected: number;
  showDuration: number;
  paused: boolean;
  count: number;
};

function useProvider({
  setCount,
  getSettingKey,
  togglePauseClickSlower,
}: {
  getSettingKey: () => string;
  setCount: MutableRefObject<((value: number) => void) | null>;
  togglePauseClickSlower: DebouncedFunc<
    (setState: (arg: SetStateAction<ProviderState>) => void) => void
  >;
}) {
  const durations = getDurations(getSettingKey());
  const [{ count, selected, showDuration, paused }, setState] =
    useState<ProviderState>({
      selected: 0,
      showDuration: Math.floor(durations.length / 2),
      paused: false,
      count: 0,
    });

  setCount.current = (value: number) => {
    if (value === count) {
      return;
    }
    setState((previous) => ({ ...previous, count: value }));
  };

  if (selected !== 0 && (selected < 0 || selected >= count)) {
    setState((current) => ({ ...current, selected: 0 }));
  }

  const duration = durations[showDuration];
  return {
    selected: selected,
    setSelected: (selected: number) => {
      setState((current) => ({ ...current, selected }));
    },
    goBack: () => {
      goBack(count, setState);
    },
    goForward: () => {
      goForward(count, setState);
    },
    showDuration: duration,
    paused: paused,
    getTotal: () => count,
    togglePause: () => {
      togglePauseClickSlower(setState);
    },
    nextShowDuration: () => {
      const next = showDuration + 1;
      setState((previous) => ({
        ...previous,
        showDuration: next >= durations.length ? 0 : next,
      }));
    },
  };
}

const leftTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function LeftDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <LeftDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: leftCount,
        getSettingKey: getLeftSettingKey,
        togglePauseClickSlower: leftTogglePauseClickSlower,
      })}
    >
      {children}
    </LeftDynamicItemSetPairContext.Provider>
  );
}

const rightTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function RightDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <RightDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: rightCount,
        getSettingKey: getRightSettingKey,
        togglePauseClickSlower: rightTogglePauseClickSlower,
      })}
    >
      {children}
    </RightDynamicItemSetPairContext.Provider>
  );
}
