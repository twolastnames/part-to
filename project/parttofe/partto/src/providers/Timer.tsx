import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { PropsWithChildren } from "react";
import { DateTime, getDateTime } from "../shared/dateTime";
import { Duration, getDuration } from "../shared/duration";
import { RunState, TaskDefinitionId } from "../api/sharedschemas";
import { Timer } from "../components/Timer/Timer";
import {
  addAlarmNote,
  removeNote,
} from "../components/Layout/NavigationBar/Noted/Noted";
import {
  addRunStateListener,
  removeRunStateListener,
} from "../shared/runStateMessage";

type TimerDescription = {
  component: ReactNode;
  duration: Duration;
  started: DateTime;
  message: string;
  enforced: boolean;
  task: TaskDefinitionId;
  setOnLocate: (callback: () => void) => void;
};

export type TimerType = {
  timers: { [id: string]: TimerDescription };
};

export const TimerContext = createContext<undefined | TimerType>(undefined);

type TaskKeyed = { task: TaskDefinitionId };

type TimerFactory<T> = (args: T) => {
  component: ReactNode;
  duration: Duration;
  task: TaskDefinitionId;
};

function getMappedTimers<T extends TaskKeyed>(
  getTimer: TimerFactory<T>,
  timers: Array<T> | undefined,
): { [id: string]: TimerDescription } {
  return (
    timers?.reduce(
      (current, timer) => ({
        ...current,
        [timer.task]: getTimer(timer),
      }),
      {},
    ) || ({} as { [id: TaskDefinitionId]: T })
  );
}

export function TimerProvider({ children }: PropsWithChildren) {
  const [runState, setRunState] = useState<RunState | undefined>();
  const offsets = useRef<{
    [id: string]: Duration;
  }>({});
  const onLocates = useRef<{ [task: string]: () => void }>({});
  const audio = useRef(new Audio(require("./messageAlert.mp3")));
  const timestamp = runState?.timestamp;
  const timers = useMemo(
    () => ({
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: true,
          message: "Duty Complete?",
          component: (
            <Timer
              start={started}
              duration={duration}
              adjustment={{
                offset: offsets.current[task] || getDuration(0),
                setOffset: (value: Duration) => (offsets.current[task] = value),
              }}
            />
          ),
        }),
        runState?.timers.enforced || [],
      ),
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: false,
          message: "Past Expected Time",
          component: <Timer start={started} duration={duration} />,
        }),
        runState?.timers.laxed || [],
      ),
      ...getMappedTimers(
        ({ task, till }) => ({
          task,
          started: timestamp,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration: till,
          enforced: true,
          message: "Need to Start Duty",
          component: <Timer start={timestamp} duration={till} />,
        }),
        runState?.timers.imminent || [],
      ),
    }),
    [runState, offsets, timestamp],
  );

  useEffect(() => {
    addRunStateListener(setRunState);
    return () => {
      removeRunStateListener(setRunState);
    };
  }, [setRunState]);

  useEffect(() => {
    const id = setInterval(() => {
      const now = getDateTime();
      for (const [key, timer] of Object.entries(timers)) {
        const expireAt = timer.started
          .add(offsets.current[key] || getDuration(0))
          .add(timer.duration);
        if (!timer.enforced || now.sinceEpoch() < expireAt.sinceEpoch()) {
          continue;
        }
        addAlarmNote({
          key,
          heading: "Alarm Expired",
          detail: timer.message,
          onClick: () => {
            onLocates.current[key]?.();
          },
        });
        if (!audio.current.paused) {
          return;
        }
        try {
          audio.current.play();
        } catch (NotAllowedError) {
          console.error(
            "could not start an alarm... likely because the page has not been interacted with",
          );
        }
        audio.current.loop = true;
        return;
      }
      for (const key of Object.keys(timers)) {
        removeNote(key);
      }
      audio.current.pause();
    }, 1000);
    return () => {
      clearInterval(id);
    };
  }, [offsets, runState, timers]);

  return (
    <TimerContext.Provider
      value={{
        timers,
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimerProvider({
  task,
  onLocate,
}: {
  task: TaskDefinitionId;
  onLocate?: () => void;
}) {
  const timers = useContext(TimerContext)?.timers;
  onLocate && timers?.[task]?.setOnLocate(onLocate);
  return timers?.[task]?.component || <></>;
}
