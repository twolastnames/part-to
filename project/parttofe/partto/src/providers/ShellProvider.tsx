import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
