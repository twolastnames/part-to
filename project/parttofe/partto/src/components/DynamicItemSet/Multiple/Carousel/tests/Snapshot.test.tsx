import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../../../providers/ShellProvider";
import { Carousel } from "../Carousel";
import { RightContext } from "../../../../../providers/DynamicItemSetPair";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Carousel
        context={RightContext}
        items={[
          {
            key: "9",
            detailView: <div>hello</div>,
            listView: <div>world</div>,
            itemOperations: [],
          },
          {
            key: "7",
            detailView: <div>world</div>,
            listView: <div>world</div>,
            itemOperations: [],
          },
        ]}
      />
    </ShellProvider>,
  );
  const component = screen.getByTestId("Carousel");
  expect(component).toMatchSnapshot();
});
