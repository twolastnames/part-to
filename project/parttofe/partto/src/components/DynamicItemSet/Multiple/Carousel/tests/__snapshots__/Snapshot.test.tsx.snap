// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="multiple"
  data-testid="Carousel"
>
  <div
    class="carousel"
  >
    <div
      class="carouselButton"
    >
      <div
        class="container"
      >
        <button
          aria-label="Back One"
          class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
          data-testid="Button"
          style="--ai-color: var(--mantine-color-white);"
          type="button"
        >
          <span
            class="m_8d3afb97 mantine-ActionIcon-icon"
          >
            <svg
              class="tabler-icon tabler-icon-caret-up "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 14l-6 -6l-6 6h12"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="listView"
      >
        <div>
          hello
        </div>
      </div>
    </div>
    <div
      class="carouselButton"
    >
      <div
        class="container"
      >
        <button
          aria-label="Forward One"
          class="mantine-focus-auto mantine-active button m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
          data-testid="Button"
          style="--ai-color: var(--mantine-color-white);"
          type="button"
        >
          <span
            class="m_8d3afb97 mantine-ActionIcon-icon"
          >
            <svg
              class="tabler-icon tabler-icon-caret-down "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 10l6 6l6 -6h-12"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class="iterationProgress"
    data-testid="IterationProgress"
  >
    <span>
      <span
        data-testid="Timer"
      >
        <span
          data-testid="LocalProgress"
        >
          <div
            class="m_b32e4812 mantine-RingProgress-root"
            data-testid="Progress"
            style="--rp-size: calc(4.5rem * var(--mantine-scale)); --rp-label-offset: calc(1.5rem * var(--mantine-scale));"
          >
            <svg
              class="m_d43b5134 mantine-RingProgress-svg"
            >
              <circle
                class="m_b1ca1fbf mantine-RingProgress-curve"
                cx="36"
                cy="36"
                fill="none"
                r="20.4"
                stroke-dasharray="128.14493602139692, 0.032044245066615885"
                stroke-dashoffset="0"
                stroke-linecap="butt"
                stroke-width="12"
                transform="scale(1, -1) translate(0, -72)"
              />
              <circle
                class="m_b1ca1fbf mantine-RingProgress-curve"
                cx="36"
                cy="36"
                fill="none"
                r="20.4"
                stroke-dasharray="0.032044245066615885, 128.14493602139692"
                stroke-dashoffset="128.17698026646354"
                stroke-linecap="butt"
                stroke-width="12"
                style="--curve-color: var(--secondary-color);"
              />
            </svg>
            <div
              class="m_b23f9dc4 mantine-RingProgress-label"
            >
              <div
                class="label"
              >
                1
              </div>
            </div>
          </div>
        </span>
      </span>
    </span>
    <span
      class="count"
    >
      <span
        data-testid="Count"
      >
        <span
          data-testid="LocalProgress"
        >
          <div
            class="m_b32e4812 mantine-RingProgress-root"
            data-testid="Progress"
            style="--rp-size: calc(4.5rem * var(--mantine-scale)); --rp-label-offset: calc(1.5rem * var(--mantine-scale));"
          >
            <svg
              class="m_d43b5134 mantine-RingProgress-svg"
            >
              <circle
                class="m_b1ca1fbf mantine-RingProgress-curve"
                cx="36"
                cy="36"
                fill="none"
                r="20.4"
                stroke-dasharray="64.08849013323177, 64.08849013323177"
                stroke-dashoffset="0"
                stroke-linecap="butt"
                stroke-width="12"
                transform="scale(1, -1) translate(0, -72)"
              />
              <circle
                class="m_b1ca1fbf mantine-RingProgress-curve"
                cx="36"
                cy="36"
                fill="none"
                r="20.4"
                stroke-dasharray="64.08849013323177, 64.08849013323177"
                stroke-dashoffset="128.17698026646354"
                stroke-linecap="butt"
                stroke-width="12"
                style="--curve-color: var(--secondary-color);"
              />
            </svg>
            <div
              class="m_b23f9dc4 mantine-RingProgress-label"
            >
              <div
                class="label"
              >
                2
              </div>
            </div>
          </div>
        </span>
      </span>
    </span>
    <span
      class="slider"
    >
      <div
        class="m_dd36362e mantine-Slider-root"
        style="--slider-radius: var(--mantine-radius-xl); --slider-thumb-size: calc(var(--slider-size) * 2);"
        tabindex="-1"
      >
        <div
          class="m_a8645c2 mantine-Slider-trackContainer"
        >
          <div
            class="m_c9ade57f mantine-Slider-track"
          >
            <div
              class="m_38aeed47 mantine-Slider-bar"
              style="--slider-bar-width: calc(40% + var(--slider-size)); --slider-bar-offset: calc(0% - var(--slider-size));"
            />
            <div
              aria-label=""
              aria-valuemax="5"
              aria-valuemin="0"
              aria-valuenow="2"
              class="mantine-focus-auto m_c9a9a60a mantine-Slider-thumb"
              role="slider"
              style="--slider-thumb-offset: 40%;"
              tabindex="0"
            />
            <div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 0%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                  data-filled="true"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  1 seconds
                </div>
              </div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 20%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                  data-filled="true"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  2 seconds
                </div>
              </div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 40%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                  data-filled="true"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  4 seconds
                </div>
              </div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 60%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  8 seconds
                </div>
              </div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 80%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  16 seconds
                </div>
              </div>
              <div
                class="m_b7b0423a mantine-Slider-markWrapper"
                style="--mark-offset: 100%;"
              >
                <div
                  class="m_dd33bc19 mantine-Slider-mark"
                />
                <div
                  class="m_68c77a5b mantine-Slider-markLabel"
                  style="display: none;"
                >
                  Paused
                </div>
              </div>
            </div>
          </div>
        </div>
        <input
          type="hidden"
          value="2"
        />
      </div>
    </span>
  </div>
</div>
`;
